File: frontend/app/shared/types/claim.ts
import { z } from "zod";

// Claim status enum
export const claimStatusSchema = z.enum([
  "DRAFT",
  "SUBMITTED",
  "REVIEWED",
  "APPROVED",
  "REJECTED",
]);

export type ClaimStatus = z.infer<typeof claimStatusSchema>;

// Create claim input
export const createClaimSchema = z.object({
  insurance_id: z.number().int().positive("Insurance ID is required"),
});

export type CreateClaimInput = z.infer<typeof createClaimSchema>;

// Update claim input
export const updateClaimSchema = z.object({
  first_name: z.string().max(100).optional(),
  last_name: z.string().max(100).optional(),
  email: z.string().email().optional(),
  phone_number: z.string().max(20).optional(),
  user_id_number: z.string().max(50).optional(),
  policy_number: z.string().max(50).optional(),
  policy_holder_number: z.string().max(50).optional(),
  coverage_start_date: z.string().optional(),
  coverage_end_date: z.string().optional(),
  claim_date: z.string().optional(),
  claim_type: z.string().max(100).optional(),
  description: z.string().optional(),
  claim_amount: z.number().min(0).optional(),
});

export type UpdateClaimInput = z.infer<typeof updateClaimSchema>;

// Claim review
export const claimReviewSchema = z.object({
  id: z.number(),
  claim_id: z.number(),
  verifier_id: z.number(),
  review_summary: z.string(),
  reviewed_at: z.string(),
});

export type ClaimReview = z.infer<typeof claimReviewSchema>;

// Claim approval
export const claimApprovalSchema = z.object({
  id: z.number(),
  claim_id: z.number(),
  approver_id: z.number(),
  decision: z.enum(["APPROVED", "REJECTED"]),
  reason: z.string().nullable(),
  decided_at: z.string(),
});

export type ClaimApproval = z.infer<typeof claimApprovalSchema>;

// Full claim response
export const claimSchema = z.object({
  id: z.number(),
  claim_number: z.string(),
  user_id: z.number(),
  insurance_id: z.number(),
  status: claimStatusSchema,
  first_name: z.string().nullable(),
  last_name: z.string().nullable(),
  email: z.string().nullable(),
  phone_number: z.string().nullable(),
  user_id_number: z.string().nullable(),
  policy_number: z.string().nullable(),
  policy_holder_number: z.string().nullable(),
  coverage_start_date: z.string().nullable(),
  coverage_end_date: z.string().nullable(),
  claim_date: z.string().nullable(),
  claim_type: z.string().nullable(),
  description: z.string().nullable(),
  claim_amount: z.number().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
  review: claimReviewSchema.nullable(),
  approval: claimApprovalSchema.nullable(),
});

export type Claim = z.infer<typeof claimSchema>;

// Paginated claims response
export const paginatedClaimsSchema = z.object({
  items: z.array(claimSchema),
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
  total_pages: z.number(),
});

export type PaginatedClaims = z.infer<typeof paginatedClaimsSchema>;

// Review claim input
export const reviewClaimSchema = z.object({
  summary: z.string().min(10, "Review summary must be at least 10 characters"),
});

export type ReviewClaimInput = z.infer<typeof reviewClaimSchema>;

// Approve claim input
export const approveClaimSchema = z.object({
  decision: z.enum(["APPROVED", "REJECTED"]),
  reason: z.string().optional(),
});

export type ApproveClaimInput = z.infer<typeof approveClaimSchema>;

File: backend/app/dtos/claim.py
from datetime import datetime, date
from decimal import Decimal
from enum import Enum
from pydantic import BaseModel, EmailStr, Field

from app.models.claim_insurance import ClaimStatus
from app.models.claim_approval import ApprovalDecision


# ============ Request DTOs ============

class ClaimCreate(BaseModel):
    """DTO for creating a new claim (creates DRAFT)."""
    insurance_id: int = Field(..., description="insurance id want to claim")



class ClaimUpdate(BaseModel):
    """DTO for updating a draft claim (patchable fields)."""
    # User info
    first_name: str | None = Field(None, max_length=100)
    last_name: str | None = Field(None, max_length=100)
    email: EmailStr | None = None
    phone_number: str | None = Field(None, max_length=20)
    user_id_number: str | None = Field(None, max_length=50)
    
    # Insurance info
    policy_number: str | None = Field(None, max_length=50)
    policy_holder_number: str | None = Field(None, max_length=50)
    coverage_start_date: date | None = None
    coverage_end_date: date | None = None
    
    # Claim details
    claim_date: date | None = None
    claim_type: str | None = Field(None, max_length=100)
    description: str | None = None
    claim_amount: Decimal | None = Field(None, ge=0, decimal_places=2)


class ClaimReviewCreate(BaseModel):
    """DTO for verifier to review a claim."""
    summary: str = Field(..., min_length=10, description="Review summary by verifier")


class ClaimApprovalCreate(BaseModel):
    """DTO for approver to approve/reject a claim."""
    decision: ApprovalDecision
    summary: str | None = Field(None, description="Reason for decision (required for rejection)")


# ============ Response DTOs ============

class ClaimReviewResponse(BaseModel):
    """DTO for claim review response."""
    id: int
    claim_id: int
    verifier_id: int
    review_summary: str
    reviewed_at: datetime
    
    class Config:
        from_attributes = True


class ClaimApprovalResponse(BaseModel):
    """DTO for claim approval response."""
    id: int
    claim_id: int
    approver_id: int
    decision: ApprovalDecision
    reason: str | None
    decided_at: datetime
    
    class Config:
        from_attributes = True


class ClaimResponse(BaseModel):
    """DTO for claim response."""
    id: int
    claim_number: str
    user_id: int
    insurance_id: int
    status: ClaimStatus
    
    # User info
    first_name: str | None
    last_name: str | None
    email: str | None
    phone_number: str | None
    user_id_number: str | None
    
    # Insurance info
    policy_number: str | None
    policy_holder_number: str | None
    coverage_start_date: date | None
    coverage_end_date: date | None
    
    # Claim details
    claim_date: date | None
    claim_type: str | None
    description: str | None
    claim_amount: Decimal | None
    
    created_at: datetime
    updated_at: datetime
    
    # Related data (optional)
    review: ClaimReviewResponse | None = None
    approval: ClaimApprovalResponse | None = None
    
    class Config:
        from_attributes = True


class ClaimListResponse(BaseModel):
    """DTO for paginated claim list."""
    items: list[ClaimResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class ClaimStatusFilter(str, Enum):
    """Filter options for claim status."""
    ALL = "all"
    DRAFT = "DRAFT"
    SUBMITTED = "SUBMITTED"
    REVIEWED = "REVIEWED"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
